### PR TITLE
Added Enemy AI

### PR DESCRIPTION
**Pull request by MadCoderr**
## Types of changes

What types of changes does your code introduce to Appium?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Description

Added Enemy AI, enemy will follow and be able to attack player and also take damage from player.
Will follow player when player reach certain distance and enemy will attack when reached to player at certain distance.

## Checklist

Put an `x` in the boxes that apply. 
_You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [X] The code compiles and has no errors.
- [ ] I have extended the README if necessary
- [X] All added code complies with the set guidelines.
- [X] I have added necessary documentation (if appropriate) to all added code.
- [ ] Any dependent changes have been merged and published in downstream modules
- [X] I have pulled the latest master into this branch to avoid merge conflicts.
- [ ] I have created new issues for any new bugs that were caused by this pull request.

## New bugs in this PR

Please describe new bugs that have occured in this pull request, also link the issues of said bugs.

## Fixed bugs by this PR

If any, please link the issues of fixed bugs. If none, type "None"

### CODE REVIEW

**Comment by GSerum** [comment]
@MadCoderr Could you double check the Checklist? If the code doesn't compile (which i assume it doesn't as the check isn't crossed) i won't merge any code as this would break the master branch
Also please do add documentation wherever possible to ensure everyone else is able to read your code

File: Cows-HacknSlash/Assets/NavMeshComponents/Editor/NavMeshComponentsGUIUtility.cs [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+    public static class NavMeshComponentsGUIUtility
+    {
+        public static void AreaPopup(string labelName, SerializedProperty areaProperty)
+        {
+            var areaIndex = -1;
+            var areaNames = GameObjectUtility.GetNavMeshAreaNames();
+            for (var i = 0; i < areaNames.Length; i++)
+            {
+                var areaValue = GameObjectUtility.GetNavMeshAreaFromName(areaNames[i]);
+                if (areaValue == areaProperty.intValue)
```
Comment (GSerum) [inline Cows-HacknSlash/Assets/NavMeshComponents/Editor/NavMeshComponentsGUIUtility.cs]: Please add curly braces to this if statement. We don't want to copy Apple's goto fail error!

File: Cows-HacknSlash/Assets/NavMeshComponents/Editor/NavMeshComponentsGUIUtility.cs [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+
+            var rect = EditorGUILayout.GetControlRect(true, EditorGUIUtility.singleLineHeight);
+            EditorGUI.BeginProperty(rect, GUIContent.none, areaProperty);
+
+            EditorGUI.BeginChangeCheck();
+            areaIndex = EditorGUI.Popup(rect, labelName, areaIndex, areaNames);
+
+            if (EditorGUI.EndChangeCheck())
+            {
+                if (areaIndex >= 0 && areaIndex < areaNames.Length - 2)
```
Comment (GSerum) [inline Cows-HacknSlash/Assets/NavMeshComponents/Editor/NavMeshComponentsGUIUtility.cs]: Same here, please add curly braces

File: Cows-HacknSlash/Assets/NavMeshComponents/Editor/NavMeshComponentsGUIUtility.cs [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+                    areaProperty.intValue = GameObjectUtility.GetNavMeshAreaFromName(areaNames[areaIndex]);
+                else if (areaIndex == areaNames.Length - 1)
```
Comment (GSerum) [inline Cows-HacknSlash/Assets/NavMeshComponents/Editor/NavMeshComponentsGUIUtility.cs]: Here too

File: Cows-HacknSlash/Assets/NavMeshComponents/Editor/NavMeshComponentsGUIUtility.cs [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+        {
+            var index = -1;
+            var count = NavMesh.GetSettingsCount();
+            var agentTypeNames = new string[count + 2];
+            for (var i = 0; i < count; i++)
+            {
+                var id = NavMesh.GetSettingsByIndex(i).agentTypeID;
+                var name = NavMesh.GetSettingsNameFromID(id);
+                agentTypeNames[i] = name;
+                if (id == agentTypeID.intValue)
```
Comment (GSerum) [inline Cows-HacknSlash/Assets/NavMeshComponents/Editor/NavMeshComponentsGUIUtility.cs]: { }

File: Cows-HacknSlash/Assets/NavMeshComponents/Editor/NavMeshComponentsGUIUtility.cs [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+        // Agent mask is a set (internally array/list) of agentTypeIDs.
+        // It is used to describe which agents modifiers apply to.
+        // There is a special case of "None" which is an empty array.
+        // There is a special case of "All" which is an array of length 1, and value of -1.
+        public static void AgentMaskPopup(string labelName, SerializedProperty agentMask)
+        {
+            // Contents of the dropdown box.
+            string popupContent = "";
+
+            if (agentMask.hasMultipleDifferentValues)
```
Comment (GSerum) [inline Cows-HacknSlash/Assets/NavMeshComponents/Editor/NavMeshComponentsGUIUtility.cs]: { }

File: Cows-HacknSlash/Assets/NavMeshComponents/Editor/NavMeshComponentsGUIUtility.cs [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+                popupContent = "\u2014";
+            else
```
Comment (GSerum) [inline Cows-HacknSlash/Assets/NavMeshComponents/Editor/NavMeshComponentsGUIUtility.cs]: { }

File: Cows-HacknSlash/Assets/NavMeshComponents/Editor/NavMeshComponentsGUIUtility.cs [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+            var agentMask = (SerializedProperty)data;
+            agentMask.ClearArray();
+            agentMask.InsertArrayElementAtIndex(0);
+            agentMask.GetArrayElementAtIndex(0).intValue = -1;
+            agentMask.serializedObject.ApplyModifiedProperties();
+        }
+
+        static string GetAgentMaskLabelName(SerializedProperty agentMask)
+        {
+            if (agentMask.arraySize == 0)
```
Comment (GSerum) [inline Cows-HacknSlash/Assets/NavMeshComponents/Editor/NavMeshComponentsGUIUtility.cs]: { }

File: Cows-HacknSlash/Assets/NavMeshComponents/Editor/NavMeshComponentsGUIUtility.cs [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+                return "None";
+
+            if (IsAll(agentMask))
```
Comment (GSerum) [inline Cows-HacknSlash/Assets/NavMeshComponents/Editor/NavMeshComponentsGUIUtility.cs]: { }

File: Cows-HacknSlash/Assets/NavMeshComponents/Editor/NavMeshComponentsGUIUtility.cs [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+
+            return "Mixed...";
+        }
+
+        static bool AgentMaskHasSelectedAgentTypeID(SerializedProperty agentMask, int agentTypeID)
+        {
+            for (var j = 0; j < agentMask.arraySize; j++)
+            {
+                var elem = agentMask.GetArrayElementAtIndex(j);
+                if (elem.intValue == agentTypeID)
```
Comment (GSerum) [inline Cows-HacknSlash/Assets/NavMeshComponents/Editor/NavMeshComponentsGUIUtility.cs]: { }

File: Cows-HacknSlash/Assets/NavMeshComponents/Editor/NavMeshComponentsGUIUtility.cs [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+            {
+                var labelName = "";
+                for (var j = 0; j < agentMask.arraySize; j++)
+                {
+                    var elem = agentMask.GetArrayElementAtIndex(j);
+                    var settingsName = NavMesh.GetSettingsNameFromID(elem.intValue);
+                    if (string.IsNullOrEmpty(settingsName))
+                        continue;
+
+                    if (labelName.Length > 0)
```
Comment (GSerum) [inline Cows-HacknSlash/Assets/NavMeshComponents/Editor/NavMeshComponentsGUIUtility.cs]: { }

File: Cows-HacknSlash/Assets/NavMeshComponents/Editor/NavMeshComponentsGUIUtility.cs (hunk 1 above)
Comment (MadCoderr) [reply Cows-HacknSlash/Assets/NavMeshComponents/Editor/NavMeshComponentsGUIUtility.cs]: well i wouldn't dare to mess around with this code as these are built in library from unity

File: Cows-HacknSlash/Assets/Scripts/PlayerController/CharController.cs [hunk 11]
```diff
@@ -57,4 +57,15 @@ public class CharController : MonoBehaviour {
...
 	}
+
+    private void OnTriggerEnter(Collider other) {
+        if (other.tag == "Enemy_Sword") {
+            this.gameObject.SetActive(false);
+            //DestoryPlayer();
+        }
+    }
+
+    private void DestoryPlayer() {
```
Comment (GSerum) [inline Cows-HacknSlash/Assets/Scripts/PlayerController/CharController.cs]: Typo, should be `DestroyPlayer()`

**Pull request merged**